### PR TITLE
use tsc typechecking for carousel 0.1

### PR DESCRIPTION
**Pull request by samouri**
**summary**
- Adds typing for BaseElement in `src/base-element.d.ts`. There are a few types I've created in this file that will need to be moved shortly, but I'd prefer to find the right homes for them in followup PRs (unless others have good ideas for where to put Layout / LayoutSize etc. right now)
- Creates a new `tsconfig` for the the carousel directory. We explicitly don't provide any `paths` so that no modules are discoverable by tsc. That way we can stub everything out. Notably, all of core stubs should go away the moment we switch it to TS. I put the tsconfig in the root of the extension because I plan on using the same config for 0.2 as well.
- TS discovered a bug: `BaseCarousel` had a private var that was being overwritten by subclasses. Removing the `observeIntersection` from the parent should be fine since the children observe separately and call super.viewportCallback.

### CODE REVIEW

File: src/base-element.d.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    user(): any;
+    getApi(): Promise<Object>;
+  }
+
+  const extension: any;
+  const registerElement: any;
+}
+
+// TODO: Move the types below to better locations.
+// Types below this line should be eventually be moved elsewhere.
```
Comment (samouri) [inline src/base-element.d.ts:102]:
```suggestion
```

File: tsconfig.json [hunk 2]
```diff
@@ -5,6 +5,7 @@
     "outDir": "dist",
     "noEmit": true,
     "paths": {
+      "#src/*": ["./src/*"],
```
Comment (samouri) [inline tsconfig.json:8]:
```suggestion
```

File: extensions/amp-carousel/0.1/slidescroll.js [hunk 3]
```diff
@@ -404,7 +404,7 @@ export class AmpSlideScroll extends BaseCarousel {
   }
 
   /**
-   * @param {!../layout-rect.LayoutSizeDef} size
+   * @param {LayoutSize} size
```
Comment (rcebulko) [inline extensions/amp-carousel/0.1/slidescroll.js:407]:
```suggestion
   * @param {!LayoutSize} size
```

File: extensions/amp-carousel/0.1/slidescroll.js [hunk 4]
```diff
@@ -999,7 +999,7 @@ export class AmpSlideScroll extends BaseCarousel {
 
   /**
    * @param {string} eventType
-   * @param {!JsonObject} vars A map of vars and their values.
+   * @param {!Object} vars A map of vars and their values.
```
Comment (rcebulko) [inline extensions/amp-carousel/0.1/slidescroll.js:1002]: Why does this change?

File: extensions/amp-carousel/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,10 @@
+{
+  "compilerOptions": {
+    "allowJs": true,
+    "checkJs": true,
+    "noEmit": true,
+    "paths": {}
+  },
+  "include": ["0.1/*.js", "0.1/*.d.ts", "../../src/*.d.ts"],
```
Comment (rcebulko) [inline extensions/amp-carousel/tsconfig.json:8]: Let's move this file into `0.1` and make each of these `**/*.js` and `**/*.d.ts`, that way the implication is "if a directory has a `tsconfig.js`, it is type-checked" (except root)

File: build-system/tasks/check-types.js [hunk 6]
```diff
@@ -48,6 +48,9 @@ const TSC_TYPECHECK_TARGETS = {
   'compiler': {
     tsconfig: 'src/compiler/tsconfig.json',
   },
+  'carousel': {
+    tsconfig: 'extensions/amp-carousel/tsconfig.json',
```
Comment (rcebulko) [inline build-system/tasks/check-types.js:52]: Hot take: Could `TSC_TYPECHECK_TARGETS` just be a list of directories which contain a `tsconfig.json`?

Comment (samouri) [reply build-system/tasks/check-types.js:52]: It could be!

File: extensions/amp-carousel/0.1/slidescroll.js (hunk 4 above)
Comment (samouri) [reply extensions/amp-carousel/0.1/slidescroll.js:1002]: JsonObject doesn't have a TS equivalent afaik. cc @jridgewell

File: extensions/amp-carousel/tsconfig.json (hunk 5 above)
Comment (samouri) [reply extensions/amp-carousel/tsconfig.json:8]: sure

File: build-system/tasks/check-types.js (hunk 6 above)
Comment (samouri) [reply build-system/tasks/check-types.js]: Done

File: extensions/amp-carousel/0.1/slidescroll.js (hunk 4 above)
Comment (jridgewell) [reply extensions/amp-carousel/0.1/slidescroll.js:1002]: JSON object is equivalent to an `any`. But, we can [finally type](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#more-recursive-type-aliases) this, if we want:

```ts
type Json =
  | string
  | number
  | boolean
  | null
  | JsonObject
  | JsonArray;
type JsonObject = { [property: string]: Json };
type JsonArray = Json[];
```

**Review by jridgewell — APPROVED** [review]

**Comment by samouri** [comment]
@rcebulko: owners approval for `build-system/check-types`

File: extensions/amp-carousel/0.1/slidescroll.js (hunk 4 above)
Comment (rcebulko) [reply extensions/amp-carousel/0.1/slidescroll.js:1002]: ^^^ @samouri can you either add this, or add an alias for `JsonObject=any` in a shame file so we can leave the annotations here and narrow them later?

**Pull request merged**